lib/tasks/multitenant/tenants.rake

```diff
@@ -68,6 +68,37 @@ namespace :multitenant do
       update_tenant_ids(proc { |object| object.account.tenant_id }, proc { account }, proc { tenant_id == nil }, args.to_hash.merge({ table_name: 'Alert' }))
     end
 
+    desc 'Check and remove orphaned objects (whose tenant is missing), pass "destroy" argument to delete'
+    task :cleanup_orphans, [:mode] => :environment do |_task, args|
+      puts 'Checking and removing orphaned objects...'
+
+      destroy = args[:mode] == "destroy"
+
+      puts "WARNING: the found orphan objects will be destroyed" if destroy
+
+      provider_account_ids = Account.where(provider: true).pluck(:id) + [Account.master.id]
+
+      ApplicationRecord.descendants.each do |model|
+        next unless model.table_exists? && model.column_names.include?('tenant_id')
+
+        orphaned_objects = model.where.not(tenant_id: provider_account_ids)
+
+        if orphaned_objects.exists?
+          puts "Found orphaned objects in #{model.table_name}:"
+          orphaned_objects.find_each { |obj| puts "- ID: #{obj.id}, Tenant ID: #{obj.tenant_id}" }
+
+          if destroy
+            puts "Destroying orphan #{model.table_name}..."
+            orphaned_objects.in_batches(of: 100).destroy_all
```

Comment:
Added batched deletion, but still not sure if it still may have negative effects on the DB...

Reply:
we need to revisit how we iterate over many objects

Reply:
> Added batched deletion, but still not sure if it still may have negative effects on the DB...

For sure batching is an improvement, what else could we do? We should be sure to run this after the billing jobs have finished and heads up ops.

Reply:
Ideally we would use the new Sidekiq 7.3 iteration jobs. But if we just do in a rake task in the foreground, batches would be better.

Reply:
If we want to run thins in produciton, we most likely need to use this iteration sidekiq job type. Because running haevy deletion in production may overburden the database. While recent hardware is much faster, I'm not sure how good this would be.